crates/accelerate/src/high_level_synthesis.rs

```diff
@@ -0,0 +1,252 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use hashbrown::HashMap;
+use pyo3::prelude::*;
+
+/// Track global qubits by their state.
+/// The global qubits are numbered by consecutive integers starting at `0`,
+/// and the states are distinguished into clean (:math:`|0\rangle`)
+/// and dirty (unknown).
+#[pyclass]
+#[derive(Clone, Debug)]
+pub struct QubitTracker {
+    /// The total number of global qubits
+    num_qubits: usize,
+    /// Stores the state for each qubit: `true` means clean, `false` means dirty
+    state: Vec<bool>,
+    /// Stores whether qubits are allowed be used
+    enabled: Vec<bool>,
+    /// Used internally for keeping the computations in `O(n)`
+    ignored: Vec<bool>,
+}
+
+#[pymethods]
+impl QubitTracker {
+    #[new]
+    pub fn new(num_qubits: usize) -> Self {
+        QubitTracker {
+            num_qubits,
+            state: vec![false; num_qubits],
+            enabled: vec![true; num_qubits],
+            ignored: vec![false; num_qubits],
+        }
+    }
+
+    /// Sets state of the given qubits to dirty
+    fn set_dirty(&mut self, qubits: Vec<usize>) {
+        for q in qubits {
+            self.state[q] = false;
+        }
+    }
```

Comment:
I was initially thinking of changing the types here to `u32` instead of `usize`, however since we mostly use the types here for indexing, I believe `usize` should suffice. We'd have to convert to `usize` anyway if we wanted to index. No action required here :)